File: gin-go-backend/internal/errors/app_error.go
package errors

type AppError struct {
	Type   string
	Title  string
	Status int
	Detail string
}

func (e AppError) Error() string { return e.Detail }

func ValidationError(detail string) AppError {
	return AppError{Type: "validation_error", Title: "Validation Error", Status: 400, Detail: detail}
}
func AuthenticationError(detail string) AppError {
	return AppError{Type: "authentication_error", Title: "Authentication Error", Status: 401, Detail: detail}
}
func AuthorizationError(detail string) AppError {
	return AppError{Type: "authorization_error", Title: "Authorization Error", Status: 403, Detail: detail}
}
func NotFoundError(detail string) AppError {
	return AppError{Type: "not_found", Title: "Not Found", Status: 404, Detail: detail}
}
func ConflictError(detail string) AppError {
	return AppError{Type: "conflict", Title: "Conflict", Status: 409, Detail: detail}
}
func RateLimitExceededError(detail string) AppError {
	return AppError{Type: "rate_limited", Title: "Too Many Requests", Status: 429, Detail: detail}
}
func ExternalServiceError(detail string) AppError {
	return AppError{Type: "external_service_error", Title: "Bad Gateway", Status: 502, Detail: detail}
}
func ServiceUnavailableError(detail string) AppError {
	return AppError{Type: "service_unavailable", Title: "Service Unavailable", Status: 503, Detail: detail}
}
func InternalServerError(detail string) AppError {
	return AppError{Type: "internal_server_error", Title: "Internal Server Error", Status: 500, Detail: detail}
}

File: actix-rust-backend/src/errors/app_error.rs
use actix_web::{http::StatusCode, HttpResponse, ResponseError};
use thiserror::Error;

use crate::errors::http_errors::ProblemDetails;

#[derive(Error, Debug)]
pub enum AppError {
    #[error("resource not found")]
    NotFound,
    #[error("service not ready")]
    ServiceUnavailable,
    #[error("rate limit exceeded")]
    RateLimited,
    #[error("internal server error")]
    Internal,
}

impl AppError {
    pub fn problem(&self, trace_id: String, instance: String) -> ProblemDetails {
        let (status, title, kind, detail) = match self {
            AppError::NotFound => (
                404,
                "Not Found",
                "https://httpstatuses.com/404",
                self.to_string(),
            ),
            AppError::ServiceUnavailable => (
                503,
                "Service Unavailable",
                "https://httpstatuses.com/503",
                self.to_string(),
            ),
            AppError::RateLimited => (
                429,
                "Too Many Requests",
                "https://httpstatuses.com/429",
                self.to_string(),
            ),
            AppError::Internal => (
                500,
                "Internal Server Error",
                "https://httpstatuses.com/500",
                self.to_string(),
            ),
        };
        ProblemDetails {
            kind: kind.into(),
            title: title.into(),
            status,
            detail,
            instance,
            trace_id,
        }
    }
}

impl ResponseError for AppError {
    fn status_code(&self) -> StatusCode {
        match self {
            AppError::NotFound => StatusCode::NOT_FOUND,
            AppError::ServiceUnavailable => StatusCode::SERVICE_UNAVAILABLE,
            AppError::RateLimited => StatusCode::TOO_MANY_REQUESTS,
            AppError::Internal => StatusCode::INTERNAL_SERVER_ERROR,
        }
    }

    fn error_response(&self) -> HttpResponse {
        HttpResponse::build(self.status_code()).finish()
    }
}
